Bluetooth: services: Call OTS selected callback for first object added

asbjornsabo opened the pull request:
Call the OTS selected callback for the first object added, as the
first object added also becomes the selected object.

This solves a startup issue where the selected callback is not called
if the first object selected by the client happens to be the same
object as the object that was first added.  In that case, the user of
the OTS does not know which object is selected, and therefore may not
be able to supply the correct data later.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

This is an alternative to https://github.com/zephyrproject-rtos/zephyr/pull/34333 - merge only one of the.

Thalley commented on subsys/bluetooth/services/ots/ots.c:
```diff
@@ -205,6 +205,17 @@ int bt_ots_obj_add(struct bt_ots *ots,
 	if (!ots->cur_obj) {
 		ots->cur_obj = obj;
+
+		char id[BT_OTS_OBJ_ID_STR_LEN];
```
Declare variables at top of scope

asbjornsabo replied:
(No longer relevant)

Thalley commented on subsys/bluetooth/services/ots/ots.c:
```diff
@@ -205,6 +205,17 @@ int bt_ots_obj_add(struct bt_ots *ots,
 	/* Make object the Current Object if this is the first one added. */
 	if (!ots->cur_obj) {
 		ots->cur_obj = obj;
+
+		char id[BT_OTS_OBJ_ID_STR_LEN];
+
+		bt_ots_obj_id_to_str(ots->cur_obj->id, id,
+					      sizeof(id));
+		LOG_DBG("Selecting a new Current Object with id: %s",
+			log_strdup(id));
+
+		if (ots->cb->obj_selected) {
+			ots->cb->obj_selected(ots, conn, ots->cur_obj->id);
+		}
```
This is a copy of the code in ots_olcp.c, right? 

I guess we could move this to an internal function `bt_ots_object_select` to avoid duplicating code

asbjornsabo replied:
I am almost ready with this.  A complication is that this code has a "conn" argument which is not available in the add_object code.  (There may not even be a connection yet.)
I can fix that by allowing a NULL pointer, but that is a change to the API.

Thalley replied:
As mentioned in the call, we can either
1) Allow NULL for conn in the callback (we allow NULL for conn in other callbacks, see e.g. obj_deleted)
2) Not select the first object at all (need to check for requirements). 

If we cannot not have a selected object, then we have a bug in `bt_ots_obj_delete` if the current selected object is deleted:
```
	if (ots->cur_obj == obj) {
		if (obj->state.type != BT_GATT_OTS_OBJECT_IDLE_STATE) {
			return -EBUSY;
		}
		ots->cur_obj = NULL;
	}
```

asbjornsabo replied:
I checked the spec - it does not forbid not having a current object.  There is even a specific error code for it - 0x81 - Object Not Selected.

I propose that we go for option 2 - not select an object until one is selected by the client.
This is a simple change to do, and it does not change the API.  And, except for fixing the issue, it makes no difference to the server - there are no local operations corresponding to the OLCP anyway.

asbjornsabo replied:
Changed the PR to not set the first object added as current.

asbjornsabo committed:
Bluetooth: services: Do not set current object until selected

Do not make the first object added to the object transfer server the
current (=selected object).  This leads to the server having no
current object until one is selected by the client.

This solves a startup issue where the selected callback is not called
by the server if the first object selected by the client happens to be
the same object as the object that was first added to the server.  In
that case, the user of the OTS does not know which object is selected,
and therefore may not be able to supply the correct data later.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>
